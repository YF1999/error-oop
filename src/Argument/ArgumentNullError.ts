import { ArgumentError } from './ArgumentError';

export class ArgumentNullError extends ArgumentError {}
