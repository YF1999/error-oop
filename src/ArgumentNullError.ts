import { ErrorOptions, ArgumentNullErrorMessageProps, ArgumentNullErrorProps } from './CommonTypes';
import { AbstractArgumentError } from './ArgumentError';

export abstract class AbstractArgumentNullError extends AbstractArgumentError {
    public constructor(
        props: ArgumentNullErrorProps,
        options: ErrorOptions<ArgumentNullErrorMessageProps>,
    ) {
        super(props, options);
    }
}

/**
 * Applicable when a null reference or undefined is passed to a function or a method that does not
 * accept it as a valid argument.
 */
export class ArgumentNullError extends AbstractArgumentNullError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be
     * appended.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     */
    public constructor(message: string, paramName: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param innerError The error that is the cause of the current error. Stack trace will be
     * appended.
     */
    public constructor(message: string, paramName: string, innerError: Error);

    public constructor(message: string = '', arg1?: string | Error, arg2?: Error) {
        function generateMessage(props: ArgumentNullErrorMessageProps) {
            return props.paramName
                ? `${props.message} (Parameter '${props.paramName}')`
                : props.message;
        }

        // message + innerError?
        if (arg1 === undefined || typeof arg1 !== 'string') {
            super({ message, innerError: arg1 }, { generateMessage });
        }

        // message + paramName + innerError?
        else {
            super({ message, paramName: arg1, innerError: arg2 }, { generateMessage });
        }
    }
}
