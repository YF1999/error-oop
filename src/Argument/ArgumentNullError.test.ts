import { ArgumentNullError } from './ArgumentNullError';

const message = 'error';
const paramName = 'key_parameter';

testError(
    'ArgumentNullError',
    () => {
        const err1 = new ArgumentNullError();
        const err2 = new ArgumentNullError(message);
        const err3 = new ArgumentNullError(message, err2);
        const err4 = new ArgumentNullError(message, paramName, err3);
        const err5 = new ArgumentNullError({ message, innerError: err4 });
        const err6 = new ArgumentNullError({ message, innerError: err5, paramName });
        const err7 = new ArgumentNullError({ message, innerError: err6, noMessageSuffix: true });
        const err8 = new ArgumentNullError({ message, innerError: err7, paramName, noMessageSuffix: true });

        return [err1, err2, err3, err4, err5, err6, err7, err8] as const;
    },
    (getErrors) => {
        test('property: message', () => {
            const [err1, ...errs] = getErrors();
            expect(err1.message).toBe('');
            errs.map((err) => expectSnapshot(err.message));
        });

        test('property: inner error', () => {
            const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
            expect(err1.innerError).toBeNull();
            expect(err2.innerError).toBeNull();
            expect(err3.innerError).toBe(err2);
            expect(err4.innerError).toBe(err3);
            expect(err5.innerError).toBe(err4);
            expect(err6.innerError).toBe(err5);
            expect(err7.innerError).toBe(err6);
            expect(err8.innerError).toBe(err7);
        });

        test('property: parameter name', () => {
            const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
            expect(err1.paramName).toBeNull();
            expect(err2.paramName).toBeNull();
            expect(err3.paramName).toBeNull();
            expect(err4.paramName).toBe(paramName);
            expect(err5.paramName).toBeNull();
            expect(err6.paramName).toBe(paramName);
            expect(err7.paramName).toBeNull();
            expect(err8.paramName).toBe(paramName);
        });
    },
);
