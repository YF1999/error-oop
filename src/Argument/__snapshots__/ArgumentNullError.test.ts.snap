// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`ArgumentNullError\` own keys 1`] = `
Array [
  "stack",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 2`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 3`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 4`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 5`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 6`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 7`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` own keys 8`] = `
Array [
  "stack",
  "message",
  "_paramName",
]
`;

exports[`Test \`ArgumentNullError\` property: message 1`] = `"error"`;

exports[`Test \`ArgumentNullError\` property: message 2`] = `"error"`;

exports[`Test \`ArgumentNullError\` property: message 3`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentNullError\` property: message 4`] = `"error"`;

exports[`Test \`ArgumentNullError\` property: message 5`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentNullError\` property: message 6`] = `"error"`;

exports[`Test \`ArgumentNullError\` property: message 7`] = `"error"`;
